File: app/src/main/java/me/cubixor/bettertello/FragmentSettingsControllerViewModel.java
package me.cubixor.bettertello;


import android.view.InputEvent;

import androidx.lifecycle.LiveData;
import androidx.lifecycle.MutableLiveData;
import androidx.lifecycle.ViewModel;

import java.util.HashMap;
import java.util.LinkedHashMap;
import java.util.Map;

import me.cubixor.bettertello.controller.Controller;
import me.cubixor.bettertello.controller.ControllerUtils;
import me.cubixor.bettertello.data.AppSettings;
import me.cubixor.bettertello.tello.TelloAction;
import me.cubixor.bettertello.utils.KeyCodes;

public class FragmentSettingsControllerViewModel extends ViewModel {

    private final AppSettings appSettings;
    private final MutableLiveData<HashMap<TelloAction, String>> buttonText = new MutableLiveData<>(new HashMap<>());
    private final MutableLiveData<TelloAction> currentButton = new MutableLiveData<>();
    private final MutableLiveData<Controller> selectedController = new MutableLiveData<>();

    public FragmentSettingsControllerViewModel() {
        appSettings = AppSettings.getInstance();

        selectedController.setValue(ControllerUtils.selectController());
    }


    public LiveData<HashMap<TelloAction, String>> getButtonText() {
        return buttonText;
    }

    public LiveData<TelloAction> getCurrentButton() {
        return currentButton;
    }

    public LiveData<Controller> getSelectedController() {
        return selectedController;
    }


    private void setMapping(TelloAction telloAction , int keyCode){
        updateButtonText(telloAction, KeyCodes.keyCodes[keyCode]);
        appSettings.addControllerMapping(getSelectedController().getValue(), telloAction, keyCode);
    }

    private void clearMapping(TelloAction telloAction) {
        updateButtonText(telloAction, "");
        appSettings.removeControllerMapping(getSelectedController().getValue(), telloAction);
    }

    public void resetMappings() {
        appSettings.resetMappings(getSelectedController().getValue());
        currentButton.setValue(null);
        updateButtonsTexts();
    }

    private void updateButtonText(TelloAction telloAction, String key) {
        HashMap<TelloAction, String> map = new HashMap<>(getButtonText().getValue());
        map.replace(telloAction, key);
        buttonText.setValue(map);
    }

    public void updateButtonsTexts() {
        Controller controller = getSelectedController().getValue();
        HashMap<TelloAction, String> map = new HashMap<>();

        for (TelloAction telloAction : TelloAction.values()) {
            int key = controller.getKeyByAction(telloAction);
            String text = "";
            if (key != -1) {
                text = KeyCodes.keyCodes[key];
            }

            map.put(telloAction, text);
        }

        buttonText.setValue(map);
    }

    public void selectController(int pos) {
        selectedController.setValue(appSettings.getControllers().get(pos));
        updateButtonsTexts();
    }

    public void changeCurrentButton(TelloAction telloAction) {
        TelloAction currentButtonValue = getCurrentButton().getValue();

        //Check if user clicked already selected this button
        if (currentButtonValue != null && currentButtonValue.equals(telloAction)) {

            //If true, remove that mapping
            clearMapping(telloAction);
            currentButton.setValue(null);
        } else {
            currentButton.setValue(telloAction);
        }

    }

    public Map<String, Integer> getControllersSpinner() {
        Map<String, Integer> controllers = new LinkedHashMap<>();

        for (Controller controller : appSettings.getControllers()) {
            controllers.put(controller.getName(), R.drawable.ic_round_sports_esports_24);
        }

        return controllers;
    }


    public boolean processInputEvent(int keyCode, InputEvent event) {
        TelloAction telloAction = getCurrentButton().getValue();

        //Check if any button is selected
        if (telloAction == null) {
            return true;
        }

        Controller controller = ControllerUtils.getControllerByID(event.getDevice().getDescriptor());

        //Check if selected controller has been used
        if (!controller.equals(getSelectedController().getValue())) {
            return true;
        }


        //Check if the key was already assigned to something and delete previous mapping
        if (controller.getMappings().containsKey(keyCode)) {
            clearMapping(controller.getMappings().get(keyCode));
        }

        setMapping(telloAction,keyCode);
        currentButton.setValue(null);

        return true;
    }


}

File: app/src/main/java/me/cubixor/bettertello/FragmentSettingsControllerViewModel.kt
package me.cubixor.bettertello

import android.view.InputEvent
import android.view.KeyEvent
import android.view.MotionEvent
import androidx.lifecycle.LiveData
import androidx.lifecycle.MutableLiveData
import androidx.lifecycle.ViewModel
import dagger.hilt.android.lifecycle.HiltViewModel
import me.cubixor.bettertello.controller.Controller
import me.cubixor.bettertello.controller.ControllerManager
import me.cubixor.bettertello.data.AppSettingsRepository
import me.cubixor.bettertello.tello.TelloAction
import me.cubixor.bettertello.utils.KeyCodes.keyCodes
import java.util.*
import javax.inject.Inject

@HiltViewModel
class FragmentSettingsControllerViewModel @Inject constructor(
    private val controllerManager: ControllerManager,
    private val appSettings: AppSettingsRepository
) : ViewModel() {


    private val _buttonText = MutableLiveData(EnumMap<TelloAction, String>(TelloAction::class.java))
    val buttonText: LiveData<EnumMap<TelloAction, String>> = _buttonText
    private val _currentButton = MutableLiveData<TelloAction?>()
    val currentButton: LiveData<TelloAction?> = _currentButton
    private val _selectedController = MutableLiveData(controllerManager.getFirstController())
    val selectedController: LiveData<Controller> = _selectedController


    private fun setMapping(telloAction: TelloAction, keyCode: Int) {
        updateButtonText(telloAction, keyCodes[keyCode])
        appSettings.addControllerMapping(selectedController.value!!, telloAction, keyCode)
    }

    private fun clearMapping(telloAction: TelloAction) {
        updateButtonText(telloAction, "")
        appSettings.removeControllerMapping(selectedController.value!!, telloAction)
    }

    fun resetMappings() {
        appSettings.resetMappings(selectedController.value!!)
        _currentButton.value = null
        updateButtonsTexts()
    }

    private fun updateButtonText(telloAction: TelloAction?, key: String) {
        val map = EnumMap(buttonText.value!!)
        map.replace(telloAction, key)
        _buttonText.value = map
    }

    private fun updateButtonsTexts() {
        val controller = selectedController.value
        val map = EnumMap<TelloAction, String>(TelloAction::class.java)
        for (telloAction in TelloAction.values()) {
            val key = controller!!.getKeyByAction(telloAction)
            var text = ""
            if (key != -1) {
                text = keyCodes[key]
            }
            map[telloAction] = text
        }
        _buttonText.value = map
    }

    fun selectController(pos: Int) {
        _selectedController.value = appSettings.controllers[pos]
        updateButtonsTexts()
    }

    fun changeCurrentButton(telloAction: TelloAction) {
        val currentButtonValue = currentButton.value

        //Check if user clicked already selected this button
        if (currentButtonValue != null && currentButtonValue == telloAction) {

            //If true, remove that mapping
            clearMapping(telloAction)
            _currentButton.value = null
        } else {
            _currentButton.value = telloAction
        }
    }

    val controllersSpinner: Map<String, Int>
        get() {
            val controllers: MutableMap<String, Int> = LinkedHashMap()
            for (controller in appSettings.controllers) {
                controllers[controller.name] = R.drawable.ic_round_sports_esports_24
            }
            return controllers
        }

    fun onGenericMotion(event: MotionEvent): Boolean {
        return controllerManager.onGenericMotionEvent(event, ::processInputEvent)
    }

    fun onKey(keyCode: Int, event: KeyEvent): Boolean {
        return controllerManager.onKeyDown(keyCode, event, ::processInputEvent)
    }

    private fun processInputEvent(keyCode: Int, event: InputEvent): Boolean {
        val telloAction = currentButton.value ?: return true

        //Check if any button is selected
        val controller = controllerManager.getControllerByID(event.device.descriptor)

        //Check if selected controller has been used
        if (controller == null || controller != selectedController.value) return true


        //Check if the key was already assigned to something and delete previous mapping
        if (controller.mappings.containsKey(keyCode)) {
            clearMapping(controller.mappings[keyCode]!!)
        }
        setMapping(telloAction, keyCode)
        _currentButton.value = null
        return true
    }
}